diff --git a/core/modules/views/src/Plugin/views/filter/Date.php b/core/modules/views/src/Plugin/views/filter/Date.php
index 84a1c42..616b670 100644
--- a/core/modules/views/src/Plugin/views/filter/Date.php
+++ b/core/modules/views/src/Plugin/views/filter/Date.php
@@ -125,6 +125,9 @@ public function acceptExposedInput($input) {
     $type = $this->value['type'];
     $rc = parent::acceptExposedInput($input);
 
+    // restore what got overwritten by the parent.
+    $this->value['type'] = $type;
+
     // Don't filter if value(s) are empty.
     $operators = $this->operators();
     if (!empty($this->options['expose']['use_operator']) && !empty($this->options['expose']['operator_id'])) {
@@ -145,8 +148,6 @@ public function acceptExposedInput($input) {
       }
     }
 
-    // restore what got overwritten by the parent.
-    $this->value['type'] = $type;
     return $rc;
   }
 
